analysis.ml

```diff
@@ -228,3 +229,38 @@ let dominates (instrs : segment) : pc -> pc -> bool =
   fun pc pc' ->
     let doms = dominators pc' in
     PcSet.exists ((=) pc) doms
+
+let lifetime_analysis (instrs : segment) =
+  let introduction instr = VarSet.elements (ModedVarSet.untyped (declared_vars instr)) in
+  let elimination instr = VarSet.elements (required_vars instr) in
+  scope_analysis introduction elimination instrs
+
+(* returns a 'pc -> pc set' computing the set of instructions depending on a declaration *)
+let required (instrs : segment) : pc -> PcSet.t =
+  let res = lifetime_analysis instrs in
+  fun pc ->
+    let instr = instrs.(pc) in
+    let declared = VarSet.elements (ModedVarSet.untyped (declared_vars instr)) in
+    let required_of var = VariableMap.at var (res pc) in
+    let all_requires = List.map required_of declared in
+    List.fold_left PcSet.union PcSet.empty all_requires
+
+(* returns a 'pc -> variable set' computing variables which need to be in scope
+ * Note: they might not be! *)
+let required_vars_at (seg : segment) : pc -> variable list =
+  let res = lifetime_analysis seg in
+  fun pc ->
+    let collect_key (key, value) = key in
+    let live_vars = List.map collect_key (VariableMap.bindings (res pc)) in
+    live_vars
+
+(* The same as required_vars_at but extends the required interval to
+ * merge points to conform to our scoping rules *)
+let required_merged_vars_at instrs =
+  let required_vars_at = required_vars_at instrs in
+  let merge pc cur_lifetime in_lifetime =
+    let merged = VarSet.union cur_lifetime in_lifetime in
+    if VarSet.equal cur_lifetime merged then None else Some merged
+  in
+  let update pc cur_lifetime = VarSet.of_list (required_vars_at pc) in
+  forward_analysis VarSet.empty instrs merge update
```

Comment:
I think we could define a generic `saturate_merge` operation that takes the result of a data-flow analysis and uses a forward_analysis path to merge joining paths. Then this would be `saturated_merge (required_vars_at instrs)`.